File: solutions/typescript/matching-brackets/1/matching-brackets.ts
export function isPaired(input: string): boolean {
    const stack: string[] = [];
    const bracketPairs: { [key: string]: string } = {
        ')': '(',
        '}': '{',
        ']': '[',
    };
    const openingBrackets = ['(', '{', '['];
    const closingBrackets = [')', '}', ']'];

    for (const char of input) {
        if (openingBrackets.includes(char)) {
            stack.push(char);
        } else if (closingBrackets.includes(char)) {
            if (stack.length === 0) {
                return false;
            }
            const lastOpen = stack.pop();
            if (lastOpen !== bracketPairs[char]) {
                return false;
            }
        }
    }

    return stack.length === 0;
}

File: solutions/go/matching-brackets/1/matching_brackets.go
package brackets

func Bracket(input string) bool {
	bracketPairs := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	var stack []rune
	for _, c := range input {
		if c == '(' || c == '{' || c == '[' {
			stack = append(stack, c)
			continue
		}
		if c == ')' || c == '}' || c == ']' {
			if len(stack) == 0 {
				return false
			}
			i := len(stack) - 1
			lastOpen := stack[i]
			stack = stack[:i]
			if lastOpen != bracketPairs[c] {
				return false
			}
		}
	}
	return len(stack) == 0
}
